Makefile

```diff
@@ -1,103 +1,187 @@
+	    fi; \
+	    scripts/build/docker-build $$buildopts $$dir
 
-# build the given commands
-build: $(PROTOCODE) $(COMMANDS)
+#
+# Targets for format checking and various code quality and complexity checks.
+#
 
 format:
-	@report=`$(GOFMT) -s -d -w $$(find cmd pkg -name \*.go)` ; if [ -n "$$report" ]; then echo "$$report"; exit 1; fi
+	@report=`$(GO_FMT) -s -d -w $$(find cmd pkg -name \*.go)`; \
+	if [ -n "$$report" ]; then \
+	    echo "$$report"; \
+	    exit 1; \
+	fi
 
 vet:
-	@$(GO) vet $(shell $(GO) list ./... | grep -v vendor)
+	@$(GO_CMD) vet $(GO_MODULES)
 
 cyclomatic-check:
-	@report=`$(GOCYCLO) -over 15 cmd pkg`; if [ -n "$$report" ]; then echo "Complexity is over 15 in"; echo $$report; exit 1; fi
+	@report=`$(GO_CYCLO) -over 15 cmd pkg`; \
+	if [ -n "$$report" ]; then \
+	    echo "Complexity is over 15 in"; \
+	    echo "$$report"; \
+	    exit 1; \
+	fi
 
 lint:
-	@rc=0 ; for f in $$(find -name \*.go | grep -v \.\/vendor) ; do $(GOLINT) -set_exit_status $$f || rc=1 ; done ; exit $$rc
+	@rc=0; \
+	for f in $$(find -name \*.go | grep -v \.\/vendor); do \
+	    $(GO_LINT) -set_exit_status $$f || rc=1; \
+	done; \
+	exit $$rc
 
 golangci-lint:
-	@$(GOLANGCI_LINT) run $(GOLANGCI_LINT_CHECKERS)
+	@$(GO_CILINT) run $(GO_CILINT_CHECKERS)
+
+
+#
+# Target for running unit/module tests.
+#
 
 test:
-	@$(GO) test -race -coverprofile=coverage.txt -covermode=atomic $(shell $(GO) list ./... | grep -v vendor)
-
-# clean the given commands
-clean:
-	@for cmd in $(COMMANDS); do \
-	    echo "Cleaning up $$cmd..." && \
-	    cd cmd/$$cmd && \
-	    go clean && \
-	    cd - > /dev/null 2>&1; \
-	done
-
-# build the given images
-images: $(DOCKERDIRS)
-	@for dd in $(DOCKERDIRS); do \
-	    echo "Building docker image for $$dd"; \
-	    buildopts="--quiet"; \
-	    if [ -n "$(DOCKERTAG)" ]; then \
-		buildopts="$$buildopts --tag $(DOCKERTAG)"; \
+	@$(GO_CMD) test -race -coverprofile=coverage.txt -covermode=atomic \
+	    $(GO_MODULES)
+
+# Target for preparing a dist-tarball for release/packaging.
+dist: git-version git-build
+	rm -fr $(TAR_DIR)* && \
+	echo "Creating tarball $(TAR_BALL)..." && \
+	git archive --format=tar --prefix=$(TAR_DIR)/ HEAD > $(TAR_BALL) && \
+	echo "Saving git version and build ids in $(TAR_BALL)..." && \
+	mkdir -p $(TAR_DIR) && cp git-version git-build $(TAR_DIR) && \
+	$(TAR_UPDATE) $(TAR_BALL) $(TAR_DIR)/git-{build,version} && \
+	echo "Compressing $(TAR_BALL) with $(TAR_COMPRESS)..." && \
+	$(TAR_COMPRESS) $(TAR_BALL) && \
+	rm -fr $(TAR_DIR)
+
+git-version:
+	@if [ -d .git ]; then \
+	    version=`git describe 2>/dev/null || echo 0.0.0unknown`; \
+	else \
+	    if [ -f git-version ]; then \
+	        version=`cat git-version`; \
+	    else \
+	        version='0.0.0unknown'; \
 	    fi; \
-	    if [ -n "$(DOCKERPUSH)" ]; then \
-	        buildopts="$$buildopts --registry $(DOCKERPUSH)"; \
+	fi; \
+	echo "$$version" > $@
+
+git-build:
+	@if [ -d .git ]; then \
+	    build=`git rev-parse --short HEAD 2>/dev/null`; \
```

Comment:
It'd be nice to see the `dirty` marker here, too(?) Just a suggestion: `git describe --always --long --dirty`

Reply:
> It'd be nice to see the `dirty` marker here, too(?) Just a suggestion: `git describe --always --long --dirty`

Hmm.. that's what I used to have there (see the PR intro examples about --version output). I wonder that the heck I force-pushed the last time.

Reply:
If the tree is dirty, maybe we should just set both of those to dirty, since there is no way knowing what was compiled into the binary... Anyway we now have the dirty-marker in the version and the buildid is just the SHA1. I'm pretty sure we'll need to tweak these later...

Reply:
I think the version/sha is userful information (both in version and build) even if the tree is dirty. Even if there is no guarantee what was built the baseline version provides some a good hint